pkgs/build-support/flatten-references-graph/src/flatten_references_graph/flatten_references_graph.py

```diff
@@ -0,0 +1,45 @@
+from toolz import curried as tlz
+
+from .lib import (
+    flatten,
+    over,
+    references_graph_to_igraph
+)
+
+from .pipe import pipe
+
+MAX_LAYERS = 127
+
+
+def create_list_of_lists_of_strings(deeply_nested_lists_or_dicts_of_graphs):
+    list_of_graphs = flatten(deeply_nested_lists_or_dicts_of_graphs)
+
+    return list(
+        filter(
+            # remove empty layers
+            lambda xs: len(xs) > 0,
+            tlz.map(
+                lambda g: g.vs["name"],
+                list_of_graphs
+            )
+        )
+    )
+
+
+def flatten_references_graph(references_graph, pipeline, exclude_paths=None):
+    if exclude_paths is not None:
+        exclude_paths = frozenset(exclude_paths)
+        references_graph = tlz.compose(
+            tlz.map(over(
+                "references",
+                lambda xs: frozenset(xs).difference(exclude_paths)
+            )),
+            tlz.remove(lambda node: node["path"] in exclude_paths)
+        )(references_graph)
```

Comment:
I haven't used `toolz` before, but even as a Haskeller I find FP idioms on Python very unreadable.

To understand this piece of code, I had to look at definition of `over`, then go to `toolz` and search for what `assoc` does. We are already on an obscure corner of Nixpkgs, three different of programming languages deep (Nix>Shell>Python), generating Docker images by hand, while doing graph algorithms, so we already are waay over our complexity budget, so preferably we shouldn't add a functional programming library on top.

I find using list comprehensions in Python goes a long way. If not, just for-loop's appending to a list. I don't know what the above code does, but is it vaguely:

```suggestion
        filtered_graph = []
        for graph in references_graph:
             if i["path"] in exclude_paths: continue
             filtered_refs = graph["references"] // exclude_paths
             filtered_graph.append(dict(**graph, references=filtered_refs))
```